### DISCUSSION TITLE
Using ORM with Postgres: "Default" field parameter in models definition has no effect

### DISCUSSION THREAD

**Discussion by nkalfa** [sqlalchemy/sqlalchemy#7276] [Usage Questions] [answered]
I am using the SQLAlchemy 1.4 ORM with postgres13 and Python 3.7

In my models.py field definition I have one field that has a default value.

**Expected behaviour:** When I run any inserts for this model, I should not have to explicitly declare a value for this field as it should be automatically populated via the default keyword.

**Actual behaviour:** The default argument has no effect. When inserting users into the database, I need to enter a specific value for the default field, or the script errors with a "init takes 'n' positional arguments, only 'x' given". The point of having the default field is to avoid having to explicitly declare a value when instantiating a class.

**Correction attempts:** I have consulted the SQLAlchemy and Postgres documentation which both show that using the default field should be trivial. I have used the arguments default, server_default, and DefaultClause. I have tried specifying text and integer types like default=text('0'). I have tried setting the default value to strings and integers. I have tried adding the declared default value to the def __init__ like self.org_num = 100

None of these have helped. Fields containing a default value continue to behave like all other fields, requiring explicit values on insert.

**base.py** (This gives me a PK ID for all columns and automatically generates the class name based on the table name.)

```
from sqlalchemy import Column, create_engine, Integer
from sqlalchemy.orm import declared_attr, declarative_base, scoped_session, sessionmaker

engine = create_engine('postgresql://user:password@localhost:5432/devdb', echo=True)

db_session = scoped_session(
    sessionmaker(
        bind=engine,
        autocommit=False,
        autoflush=False
    )
)

class Base:
    @declared_attr
    def __tablename__(cls):
        return cls.__name__.lower()

    id = Column(Integer, primary_key=True)

Base = declarative_base(cls=Base)
```

**models.py**
```
from base import Base
from sqlalchemy import Column, DateTime, Text


class User(Base):

    email = Column(Text, nullable=False, unique=True)
    name = Column(Text)
    org = Column(Text, nullable=False, index=True)
    date_last_logged = Column(DateTime(timezone=True))
    org_num = Column(Text, server_default="100")

    def __init__(self, email, name, org, date_last_logged, org_num):
        super().__init__()
        self.email = email
        self.name = name
        self.org = org
        self.date_last_logged = date_last_logged
        self.org_num = org_num

```
Example of an insert statement:

```
Base.metadata.create_all(bind=engine)

alice = User('alice@throughthelookingglass.edu', 'Alice', 'Wonderland', date(1865, 11, 26), 'a default string that should be an org_num')
db_session.add_all(alice)

db_session.commit()
db_session.close()
```

An additional confusion is the need to have a def __init__() in my models.py Class. The documentation indicates I should not need this as our models are defined via the Declarative system. Our Class gets an __init__() method constructor which automatically accepts keyword names that match the columns we’ve mapped. If I commit the empty table, I am able to do so without __init__. However as soon as I am trying to instantiate classes of User without __init__ there is an errror: "TypeError: __init__() takes 1 positional argument but 5 were given."

**Comment by CaselIT** [comment]
Hi,

The issue here is your init method. You marked `org_num` are required, since it has no default value.
The default init that the sqlalchemy provides accepts only kw arguments, so you need to use `User(email='foo@bar.com', ...)`

If you remove your init or modify it to have org_num as optional the default will take place correctly

**Reply by nkalfa** [reply]
Thanks for the swift response!

I cannot remove the init because I receive the error  "TypeError: init() takes 1 positional argument but 5 were given." (Which is odd as the documentation indicates I should not need this at all as our models are defined via the Declarative system. )

Instantiating my User instance with keyword arguments also has no effect.

Can you clarify what you mean by "You marked org_num are required, since it has no default value."?
I have it defined with a default value of 100:
`org_num = Column(Text, server_default="100")`
Even if  I hardcode into the init something like `self.test_default = 100`, the default value is still required and will only give the value explicitly passed when we instance our User.

**Comment by CaselIT** [comment, marked as answer]
> I cannot remove the init because I receive the error "TypeError: init() takes 1 positional argument but 5 were given." (Which is odd as the documentation indicates I should not need this at all as our models are defined via the Declarative system. )

As I explained above you need to use keyword arguments, not positional ones with the default init

> Can you clarify what you mean by "You marked org_num are required, since it has no default value."?

Your init is `def __init__(self, email, name, org, date_last_logged, org_num):`. orn_num has no default value in the init, so a value must be passed.
If you want to keep a custom init should assign a default value in the init to org_num. So something like (assuming null is not a valid value)
```py
def __init__(self, email, name, org, date_last_logged, org_num=None):
        ...
        if org_num is not None:
            self.org_num = org_num
```


> Instantiating my User instance with keyword arguments also has no effect.

I'm not sure what you mean here. The default init of sqlalchemy will set the attributes passed to the init as kw arguments

**Reply by nkalfa** [reply]
Thank you for this explanation! It has helped me resolve this issue.
